File: 백준/Gold/16938. 캠프 준비/캠프 준비.py
N, L, R, X = map(int,input().split())
problems = list(map(int,input().split()))
problems.sort()

def dfs(depth, idx, min_d, max_d, sum_diff):
    global answer
    if sum_diff > R:
        return 
    if depth >= 2 and L <= sum_diff <= R:
        if max_d - min_d  >= X:
            answer += 1
    
    for i in range(idx+1, N):
        if min_d == -1:
            dfs(depth+1, i, problems[i], max_d, sum_diff + problems[i])
        else:
            dfs(depth+1, i, min_d, problems[i], sum_diff + problems[i])

answer = 0
dfs(0, -1, -1, -1, 0)
print(answer)

File: 백준/Gold/16938. 캠프 준비/캠프 준비.java
import java.io.*;
import java.util.*;

class Main {
  static int N,L,R,X;
  static int[] problems;
  static int answer; 

  public static void main(String args[]) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());

    N = Integer.parseInt(st.nextToken());
    L = Integer.parseInt(st.nextToken());
    R = Integer.parseInt(st.nextToken());
    X = Integer.parseInt(st.nextToken());

    problems = new int[N];
    
    st = new StringTokenizer(br.readLine());
    for (int i = 0; i < N; i++){
      problems[i] = Integer.parseInt(st.nextToken());
    }

    Arrays.sort(problems);

    dfs(0,-1,-1,-1,0);
    System.out.println(answer);
  }

  static void dfs(int depth, int idx, int minDifficult, int maxDifficult, int difficultSum) {
    if (difficultSum > R) {
      return ;
    }
    if ( L <= difficultSum && R >= difficultSum && depth >= 2 && maxDifficult - minDifficult >= X){
      answer++;
    }

    for (int i = idx+1; i < N; i++) {
        if (minDifficult == -1){
          dfs(depth+1, i, problems[i], maxDifficult, difficultSum + problems[i]);
        } else {
          dfs(depth+1, i, minDifficult, problems[i], difficultSum + problems[i]);
        }
    }
  }

}